File: 160_MatrixMedian.cpp
/*
Problem: Matrix Median
Description: Find median of a row-wise sorted matrix.

Approach: Binary search on answer. Count elements <= mid in each row. Median has more than half elements <= it.
*/

#include <bits/stdc++.h>
using namespace std;

int countSmallerEqual(vector<int>& row, int target) {
    return upper_bound(row.begin(), row.end(), target) - row.begin();
}

int matrixMedian(vector<vector<int>>& matrix) {
    int rows = matrix.size();
    int cols = matrix[0].size();
    
    int minVal = INT_MAX, maxVal = INT_MIN;
    for (int i = 0; i < rows; i++) {
        minVal = min(minVal, matrix[i][0]);
        maxVal = max(maxVal, matrix[i][cols - 1]);
    }
    
    int desired = (rows * cols + 1) / 2;
    
    while (minVal < maxVal) {
        int mid = minVal + (maxVal - minVal) / 2;
        int count = 0;
        
        for (int i = 0; i < rows; i++) {
            count += countSmallerEqual(matrix[i], mid);
        }
        
        if (count < desired) {
            minVal = mid + 1;
        } else {
            maxVal = mid;
        }
    }
    
    return minVal;
}

int main() {
    vector<vector<int>> matrix = {
        {1, 3, 5},
        {2, 6, 9},
        {3, 6, 9}
    };
    
    cout << "Median: " << matrixMedian(matrix) << endl;
    // Output: 5
    
    return 0;
}

/*
Time Complexity: O(rows * log(cols) * log(max-min))
Space Complexity: O(1)
*/

File: 160_MatrixMedian.py
"""
Problem: Matrix Median
Description: Find median of a row-wise sorted matrix.

Approach: Binary search on answer. Count elements <= mid in each row. Median has more than half elements <= it.
"""

from typing import List
import bisect

def countSmallerEqual(row: List[int], target: int) -> int:
    return bisect.bisect_right(row, target)

def matrixMedian(matrix: List[List[int]]) -> int:
    rows = len(matrix)
    cols = len(matrix[0])
    
    minVal = min(row[0] for row in matrix)
    maxVal = max(row[-1] for row in matrix)
    
    desired = (rows * cols + 1) // 2
    
    while minVal < maxVal:
        mid = (minVal + maxVal) // 2
        count = 0
        
        for row in matrix:
            count += countSmallerEqual(row, mid)
        
        if count < desired:
            minVal = mid + 1
        else:
            maxVal = mid
    
    return minVal

def main():
    matrix = [
        [1, 3, 5],
        [2, 6, 9],
        [3, 6, 9]
    ]
    
    print(f"Median: {matrixMedian(matrix)}")
    # Output: 5

if __name__ == "__main__":
    main()

"""
Time Complexity: O(rows * log(cols) * log(max-min))
Space Complexity: O(1)
"""
